include/multipass/exceptions/clone_exceptions.h

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (C) Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef MULTIPASS_CLONE_EXCEPTIONS_H
+#define MULTIPASS_CLONE_EXCEPTIONS_H
+
+#include <stdexcept>
+
+namespace multipass
+{
+class cloneInvalidNameException : public std::runtime_error
+{
+public:
+    using std::runtime_error::runtime_error;
```

Comment:
Why don't you create a string template that takes the vm name and clone name and formats them into an appropriate error string. Otherwise, there's no real point in having this class other than the class name...

Reply:
If we look at the function body of `generate_destination_instance_name_for_clone`, there are three error messages of invalid name and they all have different formats. First, I think it is informative to users to have these specific error messages. With this in place, it is hard to patternize the error message into an exception. 

The only role of this exception is to let the daemon::clone catch so he can set `grpc::StatusCode::INVALID_ARGUMENT` uniformly for these three cases. If we do not use an exception here then `generate_destination_instance_name_for_clone` needs to have the error string, which will make the function signature more cumbersome. 

`std::expected` from C++23 which is similar to Rust `std::result` will be a better choice here than an exception. But we do not have it.

Reply:
Ok, this is up for discussion, but if it were me I would probably inline `generate_destination_instance_name_for_clone()` and return a `GRPC::Status` with an appropriate message for each instance of throwing a this exception. The way I see it, the only purpose of this class is as a fancy JUMP statement. Otherwise, it provides no value...

Reply:
I think it is necessary to have `generate_destination_instance_name_for_clone` code as a function in terms of modulization, expressiveness and not cluttering the `daemon::clone` further. 

With that said, throwing an exception is one way and returning the error message from the function is another way to propagate the error message. Returning the error message requires changing the return type from `std::string` to either a struct or `std::pair`. 

> The way I see it, the only purpose of this class is as a fancy JUMP statement. Otherwise, it provides no value...

That is right. 

@ricab in case you have any better idea.